File: workers/src/game/bot/card-tracker.ts
import { decodeCard } from '../card';
import { buildFourPlayerDeck } from '../deck';
import type { SuitName, TrickPlay } from '../types';
import { RANK_VALUES } from '../types';

export class CardTracker {
  private played = new Set<string>();
  private _knownVoids = new Map<number, Set<SuitName>>();

  recordTrick(plays: TrickPlay[]): void {
    for (const play of plays) this.played.add(play.card);
  }

  recordPlay(_seat: number, card: string): void {
    this.played.add(card);
  }

  inferVoid(seat: number, suit: SuitName): void {
    if (!this._knownVoids.has(seat)) this._knownVoids.set(seat, new Set());
    this._knownVoids.get(seat)!.add(suit);
  }

  get knownVoids(): Map<number, Set<SuitName>> {
    return this._knownVoids;
  }

  remainingCards(myHand: string[]): string[] {
    const allCodes = buildFourPlayerDeck().map(c => c.code);
    const held = new Set(myHand);
    return allCodes.filter(c => !this.played.has(c) && !held.has(c));
  }

  isHighestRemaining(card: string, myHand: string[]): boolean {
    const decoded = decodeCard(card);
    if (decoded.isJoker) return true;
    const suit = decoded.suit!;
    const value = RANK_VALUES[decoded.rank!];
    const remaining = this.remainingCards(myHand);
    for (const c of remaining) {
      const d = decodeCard(c);
      if (!d.isJoker && d.suit === suit && RANK_VALUES[d.rank!] > value) return false;
    }
    return true;
  }

  isSuitExhausted(suit: SuitName, myHand: string[]): boolean {
    return !this.remainingCards(myHand).some(c => {
      const d = decodeCard(c);
      return !d.isJoker && d.suit === suit;
    });
  }

  trumpsRemaining(trumpSuit: SuitName, myHand: string[]): number {
    return this.remainingCards(myHand).filter(c => {
      const d = decodeCard(c);
      return !d.isJoker && d.suit === trumpSuit;
    }).length;
  }

  reset(): void {
    this.played.clear();
    this._knownVoids.clear();
  }
}

File: lib/offline/bot/card_tracker.dart
import 'package:koutbh/shared/models/card.dart';

class CardTracker {
  final Set<GameCard> _played = {};
  final Map<int, Set<Suit>> _knownVoids = {};

  void recordPlay(int seat, GameCard card) {
    _played.add(card);
  }

  void inferVoid(int seat, Suit suit) {
    _knownVoids.putIfAbsent(seat, () => {}).add(suit);
  }

  Set<GameCard> get playedCards => Set.unmodifiable(_played);

  /// All cards NOT yet played and NOT in my hand.
  Set<GameCard> remainingCards(List<GameCard> myHand) {
    return GameCard.fullDeck().difference(_played).difference(myHand.toSet());
  }

  Map<int, Set<Suit>> get knownVoids => Map.unmodifiable(_knownVoids);

  int trumpsRemaining(Suit trumpSuit, List<GameCard> myHand) {
    return remainingCards(
      myHand,
    ).where((c) => !c.isJoker && c.suit == trumpSuit).length;
  }

  bool isHighestRemaining(GameCard card, List<GameCard> myHand) {
    if (card.isJoker) return true;
    final suit = card.suit!;
    final remaining = remainingCards(
      myHand,
    ).where((c) => !c.isJoker && c.suit == suit);
    if (remaining.isEmpty) return true;
    final highestRemaining = remaining
        .map((c) => c.rank!.value)
        .reduce((a, b) => a > b ? a : b);
    return card.rank!.value > highestRemaining;
  }

  bool isSuitExhausted(Suit suit, List<GameCard> myHand) {
    return remainingCards(
      myHand,
    ).where((c) => !c.isJoker && c.suit == suit).isEmpty;
  }

  void reset() {
    _played.clear();
    _knownVoids.clear();
  }
}
